### PR TITLE
Change intial best beefy block

### PR DESCRIPTION
**Pull request by Wizdave97**
This PR provides a runtime storage value in the session pallet that holds the block number of when session changes occur. 

Fixes https://github.com/paritytech/grandpa-bridge-gadget/issues/256 when done

polkadot companion: https://github.com/paritytech/polkadot/pull/4729

### CODE REVIEW

File: primitives/session/src/lib.rs [hunk 1]
```diff
@@ -47,6 +47,12 @@ sp_api::decl_runtime_apis! {
 		/// Returns the list of public raw public keys + key type.
 		fn decode_session_keys(encoded: Vec<u8>) -> Option<Vec<(Vec<u8>, KeyTypeId)>>;
 	}
+
+	/// Sesson boundary runtime api
+	pub trait SessionBoundaryApi<BlockNumber: codec::Codec> {
+		/// Returns the block number at which the current session began
+		fn get_session_boundary() -> BlockNumber;
+	}
```
Comment (bkchr) [inline primitives/session/src/lib.rs:55]: Why is that its own runtime api?

IMO it should just be added to the Beefy runtime api.

File: frame/session/src/lib.rs [hunk 2]
```diff
@@ -955,3 +967,9 @@ impl<T: Config, Inner: FindAuthor<u32>> FindAuthor<T::ValidatorId>
 		validators.get(i as usize).map(|k| k.clone())
 	}
 }
+
+impl<T: Config> SessionBoundary<T::BlockNumber> for Pallet<T> {
+	fn get_session_boundary() -> T::BlockNumber {
+		Pallet::<T>::session_start()
+	}
+}
```
Comment (bkchr) [inline frame/session/src/lib.rs:975]: Why do we need an extra trait here?

File: client/beefy/src/worker.rs [hunk 3]
```diff
@@ -247,14 +247,19 @@ where
...
 
-				// this metric is kind of 'fake'. Best BEEFY block should only be updated once we
-				// have a signed commitment for the block. Remove once the above TODO is done.
-				metric_set!(self, beefy_best_block, *notification.header.number());
+				// this metric is kind of 'fake'. Best BEEFY block should only be updated once
+				// we have a signed commitment for the block. Remove once the above TODO is
+				// done.
+				if session_boundary.is_some() {
+					metric_set!(self, beefy_best_block, session_boundary.unwrap());
+				}
```
Comment (bkchr) [inline client/beefy/src/worker.rs:262]:
```suggestion
				if let Some(session_boundary) = session_boundary {
					metric_set!(self, beefy_best_block, session_boundary);
				}
```

**Review by bkchr — CHANGES_REQUESTED** [review]

**Comment by Wizdave97** [comment]
The polkadot companion has been fixed, the CI checks need to be restarted

**Review by acatangiu — APPROVED** [review]
Looks good!

**Comment by acatangiu** [comment]
@bkchr please take a look at this PR and its companion, thanks!

**Comment by seunlanlege** [comment]
Bump

**Comment by acatangiu** [comment]
Alternative PR https://github.com/paritytech/substrate/pull/10882 merged

**Pull request closed without merging**